File: libs/evals/src/aap/evals/eval/judge.py
"""LLM-as-judge scoring — evaluates whether an edit instruction was fulfilled."""

from __future__ import annotations

import asyncio
from pathlib import Path

from pydantic import BaseModel, Field
from pydantic_ai import Agent
from pydantic_ai.models import Model

from ..models import JudgeScore, TurnJudgeComparison
from .metrics import strip_aap_markers

_JUDGE_SYSTEM = """\
You are evaluating whether a text artifact correctly implements an edit instruction.

Given the edit instruction and the resulting artifact, score how well the edit was fulfilled.

- Score 0.0 = the edit was not applied at all
- Score 0.5 = partially applied (some elements present, others missing)
- Score 1.0 = perfectly fulfilled

Focus ONLY on whether the specific changes requested are present in the output.
Do NOT penalize for differences in style, formatting, or structure.
Do NOT penalize for the presence of XML-like markers or annotations.
"""


class JudgeOutput(BaseModel):
    """Structured output from the judge LLM."""

    score: float = Field(ge=0.0, le=1.0, description="Edit fulfillment score")
    reasoning: str = Field(description="Brief explanation of the score")


async def judge_turn(
    model: Model,
    edit_instruction: str,
    artifact_text: str,
    flow: str,
    turn: int,
) -> JudgeScore:
    """Score a single turn's output against the edit instruction."""
    agent: Agent[None, JudgeOutput] = Agent(
        model,
        system_prompt=_JUDGE_SYSTEM,
        output_type=JudgeOutput,
    )

    user_msg = (
        f"## Edit Instruction\n\n{edit_instruction}\n\n"
        f"## Resulting Artifact\n\n```\n{artifact_text}\n```"
    )

    r = await agent.run(user_msg)
    out = r.output

    return JudgeScore(
        turn=turn,
        flow=flow,
        score=round(out.score, 4),
        reasoning=out.reasoning,
    )


async def judge_experiment(
    model: Model,
    exp_dir: Path,
    ext: str,
) -> list[TurnJudgeComparison]:
    """Judge all turns for one experiment, scoring both flows concurrently."""
    base_input = exp_dir / "inputs" / "base"
    base_output = exp_dir / "outputs" / "base"
    aap_output = exp_dir / "outputs" / "aap"

    # Find edit turn files (turn-1, turn-2, ... — skip turn-0 which is creation)
    turn_files = sorted(
        base_input.glob("turn-*.md"),
        key=lambda p: int(p.stem.split("-")[1]),
    )
    edit_turns = [tf for tf in turn_files if int(tf.stem.split("-")[1]) >= 1]

    comparisons: list[TurnJudgeComparison] = []

    for tf in edit_turns:
        turn_num = int(tf.stem.split("-")[1])
        edit_instruction = tf.read_text().strip()
        turn_name = tf.stem  # e.g. "turn-1"

        base_file = base_output / f"{turn_name}{ext}"
        aap_file = aap_output / f"{turn_name}{ext}"

        if not base_file.exists() or not aap_file.exists():
            continue

        base_text = base_file.read_text()
        aap_text = strip_aap_markers(aap_file.read_text())

        # Judge base and AAP concurrently
        base_score, aap_score = await asyncio.gather(
            judge_turn(model, edit_instruction, base_text, "base", turn_num),
            judge_turn(model, edit_instruction, aap_text, "aap", turn_num),
        )

        comparisons.append(TurnJudgeComparison(
            turn=turn_num,
            edit_instruction=edit_instruction[:120],
            base_score=base_score.score,
            aap_score=aap_score.score,
            base_reasoning=base_score.reasoning,
            aap_reasoning=aap_score.reasoning,
        ))

    return comparisons

File: apps/eval-cli/scorer/judge.go
package scorer

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"

	agent "github.com/urmzd/saige/agent"
	agenteval "github.com/urmzd/saige/agent/eval"
	"github.com/urmzd/saige/agent/types"

	"github.com/urmzd/generative-artifact-protocol/eval-cli/gap"
)

const judgeSystem = `You are evaluating whether a text artifact correctly implements an edit instruction.

Given the edit instruction and the resulting artifact, score how well the edit was fulfilled.

- Score 0.0 = the edit was not applied at all
- Score 0.5 = partially applied (some elements present, others missing)
- Score 1.0 = perfectly fulfilled

Focus ONLY on whether the specific changes requested are present in the output.
Do NOT penalize for differences in style, formatting, or structure.
Do NOT penalize for the presence of XML-like markers or annotations.

Respond with a JSON object: {"score": 0.0-1.0, "reasoning": "brief explanation"}`

// JudgeOutput is the structured output from the judge LLM.
type JudgeOutput struct {
	Score     float64 `json:"score" description:"Edit fulfillment score from 0.0 to 1.0"`
	Reasoning string  `json:"reasoning" description:"Brief explanation of the score"`
}

// JudgeTurn scores a single turn's output against an edit instruction.
func JudgeTurn(ctx context.Context, provider types.Provider, editInstruction, artifactText, flow string, turn int) (gap.JudgeScore, error) {
	schema := types.SchemaFrom[JudgeOutput]()
	ag := agent.NewAgent(agent.AgentConfig{
		SystemPrompt:   judgeSystem,
		Provider:       provider,
		MaxIter:        1,
		ResponseSchema: &schema,
	})

	userMsg := fmt.Sprintf("## Edit Instruction\n\n%s\n\n## Resulting Artifact\n\n```\n%s\n```", editInstruction, artifactText)
	stream := ag.Invoke(ctx, []types.Message{types.NewUserMessage(userMsg)})
	_, rawText, _ := agenteval.CollectStreamTiming(stream.Deltas())
	if err := stream.Wait(); err != nil {
		return gap.JudgeScore{}, fmt.Errorf("judge %s turn-%d: %w", flow, turn, err)
	}

	var out JudgeOutput
	if err := json.Unmarshal([]byte(rawText), &out); err != nil {
		return gap.JudgeScore{}, fmt.Errorf("parse judge output: %w", err)
	}

	// Clamp score to [0, 1].
	if out.Score < 0 {
		out.Score = 0
	}
	if out.Score > 1 {
		out.Score = 1
	}

	return gap.JudgeScore{
		Turn:      turn,
		Flow:      flow,
		Score:     round4(out.Score),
		Reasoning: out.Reasoning,
	}, nil
}

// JudgeExperiment judges all edit turns for one experiment, scoring both flows concurrently.
func JudgeExperiment(ctx context.Context, provider types.Provider, expDir, ext string) ([]gap.TurnJudgeComparison, error) {
	baseInput := filepath.Join(expDir, "inputs", "base")
	baseOutput := filepath.Join(expDir, "outputs", "base")
	gapOutput := filepath.Join(expDir, "outputs", "gap")

	pattern := filepath.Join(baseInput, "turn-*.md")
	turnFiles, _ := filepath.Glob(pattern)
	sort.Strings(turnFiles)

	var comparisons []gap.TurnJudgeComparison
	for _, tf := range turnFiles {
		turnNum := parseTurnFromFilename(filepath.Base(tf))
		if turnNum < 1 {
			continue // skip turn-0 (creation, not edit)
		}

		turnName := fmt.Sprintf("turn-%d", turnNum)
		editInstruction, _ := os.ReadFile(tf)
		baseFile := filepath.Join(baseOutput, turnName+ext)
		gapFile := filepath.Join(gapOutput, turnName+ext)

		if _, err := os.Stat(baseFile); err != nil {
			continue
		}
		if _, err := os.Stat(gapFile); err != nil {
			continue
		}

		baseText, _ := os.ReadFile(baseFile)
		gapText := gap.StripGAPMarkers(string(must(os.ReadFile(gapFile))))

		// Judge base and GAP concurrently.
		var (
			baseScore, gapScore gap.JudgeScore
			baseErr, gapErr     error
			wg                  sync.WaitGroup
		)
		wg.Add(2)
		go func() {
			defer wg.Done()
			baseScore, baseErr = JudgeTurn(ctx, provider, string(editInstruction), string(baseText), "base", turnNum)
		}()
		go func() {
			defer wg.Done()
			gapScore, gapErr = JudgeTurn(ctx, provider, strings.TrimSpace(string(editInstruction)), gapText, "gap", turnNum)
		}()
		wg.Wait()

		if baseErr != nil || gapErr != nil {
			continue
		}

		comparisons = append(comparisons, gap.TurnJudgeComparison{
			Turn:            turnNum,
			EditInstruction: truncateStr(string(editInstruction), 120),
			BaseScore:       baseScore.Score,
			GAPScore:        gapScore.Score,
			BaseReasoning:   baseScore.Reasoning,
			GAPReasoning:    gapScore.Reasoning,
		})
	}

	return comparisons, nil
}

func must(data []byte, err error) []byte {
	if err != nil {
		return nil
	}
	return data
}

func truncateStr(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}
